### ISSUE TITLE
[Feature Request] Enhance error message & add error code for expired client cert

### ISSUE THREAD

**Issue by rohitgup14** [temporalio/features#9]
### Is your feature request related to a problem? Please describe.

During one of the issues while debugging with customer it was noticed that error message when client cert is expired is not very useful and clear. 

Current error message says `Context deadline exceeded`

### Describe the solution you'd like

Have a message which clearly states `Expired Certificate or something equivalent`. Also it will be beneficial if we can have error code associated with this message

**Comment by sergeybykov** [comment]
There are two two more related asks:

1. For SDK to auto-reload certificates from the file system. This will make it work well with the typical setup when certificates are injected via a mounted and get refreshed there by the infrastructure.
2. Produce a metric with remaining time until expiration of certificates. That will allow users to monitor and alert before a client certificate expires.

**Comment by cretz** [comment]
@sergeybykov -

1. File-system based certs are not even a part of the SDKs. In many cases, people can live generate them, load them from Vault, whatever. We shouldn't even assume disk is present if we don't have to. Rather, we may be able to provide a "certificate getter" that is auto-reinvoked or a "certificate setter" that a user is expected to reinvoke on SIGUSR1 or expiration or whatever.
2. Metrics don't really keep time remaining, and since they are numbers, we can have something like "unix timestamp" of expiration. But really, as a library, any caller can just as easily obtain this value themselves when they load their certificate.

Are there any other examples of, say, gRPC/HTTP-based client libraries (Go, TS, Java, etc) that handle certificate refresh internally we can look to for inspiration? Usually one might expect this to be a caller problem like any other HTTP client. For instance, in Go, one might expect them to provide `GetClientCertificate` on [tls.Config](https://pkg.go.dev/crypto/tls#Config). I think we can add a sample or even some helpers to help with this.

**Comment by Sushisource** [comment]
> Usually one might expect this to be a caller problem like any other HTTP client

Yeah, it seems a bit odd to me that any "long lived process with an HTTPS client" would all have to implement some kind of periodic cert refresh.

We definitely should do it if there's no other good way, but I'd hope there's some reasonable generic approach here.

**Comment by bergundy** [comment]
👍 to what @cretz said.

As a first step I'd verify that the right error is propagated from the SDK and that we don't attempt to reconnect in face of such errors. At least then the easy solution would be to crash the process so a new one can be started and certificates would be reloaded.

**Comment by cretz** [comment]
Ok, I have done research on this for Go. See the attached script at [main.go.txt](https://github.com/temporalio/sdk-features/files/7927798/main.go.txt) to replicate.

Basically I start a reverse proxy w/ a good client cert, change that to being a bad client cert, then change it back waiting a bit between all steps. What I learned:

* gRPC is built to retry on these failures and does so with no logs of its own
* We record the poll error once it has timed out, but we do it at the `DEBUG` level, and it's generic
* The client doesn't not know the reason for handshake failure, but the default Go HTTP server logs it properly (though the gRPC server may not log this properly):
> http: TLS handshake error from 127.0.0.1:63353: tls: failed to verify client certificate: x509: certificate has expired or is not yet valid: current time 2022-01-24T11:54:58-06:00 is after 2022-01-19T17:54:56
* The server just aborts the invalid handshake

The onus is really on the server to report its reasons for cert validation failure. We can bump the poll-failure SDK logs to `WARN` if desired, but they still only output one of these two lines:

> Error last connection error: connection closed before server preface received
> Error last connection error: use of closed network connection

**Comment by bergundy** [comment]
> Error last connection error: connection closed before server preface received

This might be helpful but we'd probably prefer metrics over logs

> Error last connection error: use of closed network connection

This one just sounds like an SDK bug

Does Temporal have any metrics for connection failures and more specifically TLS failures?

**Comment by cretz** [comment]
> This might be helpful but we'd probably prefer metrics over logs

We do already have `temporal_long_request_failure` which is basically this (we don't long poll for much else IIRC). But we can make a special metric for poll failures if wanted.

> This one just sounds like an SDK bug

It's not, this is how gRPC operates. It tries until timeout and reports the last error. This is unrelated to SDK.

> Does Temporal have any metrics for connection failures and more specifically TLS failures?

I do not know, but it probably should. It is the server that knows why it's closing the connection.
